Add support for Liked songs, swap to Tekore client for Spotify

CarvellWakeman opened the pull request:
Hey @couldbejake , I forked your repo to add support for downloading the user's spotify "liked songs".
I plan to run this script on a cron job to keep a synced directory of my liked songs.

Wasn't really intending to submit this as a PR, which is why it contains so many unrelated changes. Feel free to leave a big list of suggestions, incorporate what you like (if anything), or ignore this unpolished PR :)

Changes:
* Swap from http calls to Spotify API to Tekore client
    * Allows User login, persists Refresh token to file so login process only happens once.
    * Implements Spotify's Paging system so that playlists > 100 songs can be retrieved.
    * Easy access to some user info like Now Playing, Top Artists/Tracks, and Liked Songs
* Add 'liked' flag to parameters. Acts like a Playlist without a URL, and retrieves the user's liked songs.
* Add login.py, which will configure user access via spotify app.
* Skip songs that are already downloaded.
    * This does not currently account for files with the same name or renamed files. Might be worth removing.
* Add / modify some print outputs all over the place.
* Added skipped songs summary at the end of download

CarvellWakeman commented on apis/youtube.py:
```diff
@@ -36,8 +36,9 @@ def search(self, search_query, max_length, min_view_count):
-
         youtube_video = pytubeYouTube(url)
+        if youtube_video.age_restricted:
+            youtube_video.bypass_age_gate()
```
This doesn't seem to work. Providing the oauth flow to pytubeYouTube also only worked once.

Need to look into it further to see if there's a long lived auth option that can be setup before the script is run.

couldbejake replied:
@CarvellWakeman Still looking through your PR, but for the most part looks good. Just to confirm, is it still possible in addition to still download songs without being logged in?

CarvellWakeman replied:
Ah good point, no. Tekore can supports application tokens ([see guide](https://tekore.readthedocs.io/en/stable/auth_guide.html)) but I have not implemented it.

Should be trivial to support, just call `tk.request_client_token(clientId, clientSecret)` and pass the same Token object as user auth into `tk.Spotify(token)`

The hard(er) part is supporting both login types through some interface. The flask server in login.py only makes the user auth more convenient to avoid copying the redirectUri into the command line.

I'll think through it a bit.

couldbejake commented:
I get:


```
Please provide a Spotify URL (right click, share, copy link):

> https://open.spotify.com/playlist/76iluwOVwnEWQ8SxJmNmwq?si=dbe7fa70715741da

Which quality would you like? (Options: low, medium, high, or specify bitrate): high

Chosen Settings

Playlist: https://open.spotify.com/playlist/76iluwOVwnEWQ8SxJmNmwq?si=dbe7fa70715741da
Song quality / bitrate: high / 160000 bps
Traceback (most recent call last):
  File "C:\Users\jakenelson\Downloads\spotify2mp3-main\spotify2mp3-main\spotify2mp3.py", line 172, in <module>
    main(playlist=url, quality=quality, min_views=DEFAULT_MIN_VIEWS_FOR_DOWNLOAD, max_length=DEFAULT_MAX_LENGTH_FOR_DOWNLOAD)
  File "C:\Users\jakenelson\Downloads\spotify2mp3-main\spotify2mp3-main\spotify2mp3.py", line 108, in main
    downloader = SpotifyDownloader(get_bitrate_from_quality(quality), max_length, min_views)
  File "C:\Users\jakenelson\Downloads\spotify2mp3-main\spotify2mp3-main\downloader.py", line 26, in __init__
    self.spotify_client = Spotify()
  File "C:\Users\jakenelson\Downloads\spotify2mp3-main\spotify2mp3-main\apis\spotify.py", line 17, in __init__
    user_token = self.get_stored_token()
  File "C:\Users\jakenelson\Downloads\spotify2mp3-main\spotify2mp3-main\apis\spotify.py", line 27, in get_stored_token
    refreshToken) = tk.config_from_file('tekore_cfg.ini', return_refresh=True)
  File "C:\Users\jakenelson\AppData\Local\Programs\Python\Python39\lib\site-packages\tekore\_config.py", line 146, in config_from_file
    c = _read_configfile(file_path)
  File "C:\Users\jakenelson\AppData\Local\Programs\Python\Python39\lib\site-packages\tekore\_config.py", line 100, in _read_configfile
    with open(file_path, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'tekore_cfg.ini'
```

The application needs the ability to download without logging in, as well as authenticated.

Regardless, thank you for your work.

CarvellWakeman commented:
@couldbejake Alright, rewrote the login process completely. It can now do anonymous downloads or user login by the parameter `--login`.

I'm looking at the linter results once their done. Should have those done today or tomorrow.

In addition to the features in the PR, I've added these:
* Integrate login process into the command line. The first time the --login parameter is supplied, or a user-scoped request is made (e.g. download Liked songs or a Private playlist) the login setup runs. The results are persisted to a cfg file so this process only has to happen once.
* When searching for youtube videos, select the top 3 and pick the highest viewed one.
* A bunch of print statement tweaks such as an error for private playlists, etc.

couldbejake commented:
The app looks brilliant so far.

The only changes I see:
- The program prompts you to login regardless of a playlist being public or not (Is this is a localization issue?)
- The fact that we're downloading from a private playlist should be more explicit.
- I don't think flask was included in the requirements.txt file, hence throws an error.
- Would it be possible to make our own Spotify app, to prevent the user from having to create their own - this might be a limiting factor with onboarding.
- The updates you made are genuinely awesome, I was in shock just watching it run

CarvellWakeman commented:
* The program prompts you to login regardless of a playlist being public or not (Is this is a localization issue?)

Do you have an example playlist? I've seen that private playlists have &pt in the query params of the URL, but I was only checking for &. If your example public playlist has an & in the share URL that's probably why. I updated private URLs to check for "&pt"

* The fact that we're downloading from a private playlist should be more explicit.
   
I missed a check for this when running from args vs interactive. Let me know if you have something else in mind than printing "Private Playlist"
![image](https://github.com/couldbejake/spotify2mp3/assets/10020591/cb02f789-78ed-4afe-b388-16ced18e0d16)

* I don't think flask was included in the requirements.txt file, hence throws an error.

Added

* Would it be possible to make our own Spotify app, to prevent the user from having to create their own - this might be a limiting factor with onboarding.

Might be possible, I didn't see anything to help that process in Tekore library. With other services I've seen this app registration process occur between the tool and the service vendor, so the user doesn't have to perform that step. Maybe Spotify has something like that?

* The downloaded song index [10/15] goes to (count - 1), I'm wondering why, as index does not start at 0

Oops, off by 1 :)

* Let's hide the YouTube link when downloading the song - A bit of magic is OK!

Removed. I think the youtube search code needs to be improved, I kept getting bad results for slightly less popular songs, so I added the URL to troubleshoot. Might be worth adding a Verbose flag and hide info behind it.

* Does login.py need to be included in the guide if the program shows it regardless?

Removed

* The updates you made are genuinely awesome, I was in shock just watching it run

No problem! It was great to work on a lower stakes software project for a little bit. Thanks for making the project!


**Youtube search improvement suggestions:**
* Search for top N videos (configurable)
* From those top N, find the best one by:
* Which video title matches the spotify track title the closest
* Which video duration  matches the spotify track duration the closest
* Which video is most popular
* Does the video's channel name match the spotify artist name?
* Misc: age restriction blocks videos at random, the device-flow auth solution in the library only worked sometimes too.

couldbejake commented:
Brilliant, thanks @CarvellWakeman
 
It looks like we still have an issue regarding public playlists prompting auth, but it seems an easy fix.

Thank you so much for your contribution, I'm sure everyone will appreciate what you've added.

I will accept the PR tomorrow when I have a little time.

I hope we meet again!

Jake

couldbejake commented:
Sorry for the delay everyone, I need some time to re-add the ability to download public playlists, albums.
